packages/core/src/fetchProxy.ts

```diff
@@ -0,0 +1,95 @@
+import { toStackTraceString } from './errorCollection'
+import { monitor } from './internalMonitoring'
+import { computeStackTrace } from './tracekit'
+import { normalizeUrl } from './urlPolyfill'
+
+export interface FetchProxy {
+  beforeSend: (callback: (context: Partial<FetchContext>) => void) => void
+  onRequestComplete: (callback: (context: FetchContext) => void) => void
+  reset: () => void
+}
+
+export interface FetchContext {
+  method: string
+  startTime: number
+  duration: number
+  url: string
+  status: number
+  response: string
+  responseType?: string
+  [key: string]: unknown
+}
+
+let originalFetch: typeof window.fetch
+let hasBeenStarted = false
+const beforeSendCallbacks: Array<(xhr: Partial<FetchContext>) => void> = []
+const onRequestCompleteCallbacks: Array<(xhr: FetchContext) => void> = []
+
+export function startFetchProxy(): FetchProxy {
```

Comment:
The API is a bit weird, because it returns a new instance, but the state is global... maybe expose standalone functions instead?

Reply:
Would you prefer something like:
```ts
let fetchProxySingleton: FetchProxy

export function startFetchProxy(): FetchProxy {
  if (!hasBeenStarted) {
    hasBeenStarted = true
    proxyFetch()
    fetchProxySingleton = {
      beforeSend...
    }
  }
  return fetchProxySingleton
}
```

Reply:
With this, you'll still need to do `startXhrProxy().reset()` which I find unintuitive.

What about something like:

```ts
export function startFetchProxy(): FetchProxy {
  if (!hasBeenStarted) {
    hasBeenStarted = true
    proxyFetch()
  }
}

export function beforeSend(callback: (context: Partial<FetchContext>) => void) {
      beforeSendCallbacks.push(callback)
}
exporn function onRequestComplete(callback: (context: FetchContext) => void) {
      onRequestCompleteCallbacks.push(callback)
 }
export function resetFetchProxy() {
      if (hasBeenStarted) {
            hasBeenStarted = false
            beforeSendCallbacks.splice(0, beforeSendCallbacks.length)
            onRequestCompleteCallbacks.splice(0, onRequestCompleteCallbacks.length)
            window.fetch = originalFetch
      }
}
```

Reply:
I would prefer to enforce the call to start before adding callbacks, to avoid to rely on a proxy started elsewhere.

with my previous suggestion, we could have:

```ts
export function resetFetchProxy() {
  if (hasBeenStarted) {
    fetchProxySingleton.reset()
  }
}
```
wdyt?

Reply:
cf 23b5be9